backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/bll/scheduling/policyunits/CPUPolicyUnit.java

```diff
@@ -34,61 +40,103 @@ public class CPUPolicyUnit extends PolicyUnitImpl {
-                        continue;
-                    }
-                } else {
-                    int takenCpus = 0;
-                    // takenCpus are CPUs (threads), we should consider switching it to cores when necessary.
-                    takenCpus = context.getCluster().getCountThreadsAsCores() ?
-                            vdsCpuUnitPinningHelper.getDedicatedCount(vds.getId()) :
-                            vdsCpuUnitPinningHelper.countTakenCores(vds);
-                    cores = cores - takenCpus;
-                    if (numOfCpus > cores) {
-                        messageNotEnoughCores(vds, cores, vm, messages);
-                        continue;
-                    }
+
+            // when the dynamic topology is not set, the host requires more then one core per socket
+            if (anyVmWithUnsetDynamicTopology) {
+                int coresPerSocket = 0;
+
+                if (host.getCpuCores() != null && host.getCpuSockets() != null && host.getCpuSockets() != 0) {
+                    coresPerSocket = host.getCpuCores() / host.getCpuSockets();
                 }
 
+                if (coresPerSocket <= 1) {
+                    messages.addMessage(host.getId(), EngineMessage.VAR__DETAIL__NOT_ENOUGH_CORES.toString());
+                    log.debug(
+                            "Host '{}' has only one core per socket. Resize and pin requires more than one core per socket",
+                            host.getName());
+                    continue;
+                }
+            }
+
+            Map<Guid, List<VdsCpuUnit>> vmToPendingExclusiveCpuPinnings =
+                    PendingCpuPinning.collectForHost(getPendingResourceManager(), host.getId());
+            int unavailableCpus = vdsCpuUnitPinningHelper.countUnavailableCpus(
+                    host,
+                    vmToPendingExclusiveCpuPinnings,
+                    context.getCluster().getCountThreadsAsCores());
+
+            int addedExclusivelyPinnedCpus = vmGroup.stream()
+                    .filter(vm -> vm.getCpuPinningPolicy().isExclusivelyPinned())
+                    .mapToInt(vm -> vm.getNumOfCpus())
+                    .sum();
+
+            long maxPendingSharedCount = pendingResourceManager.pendingHostResources(host.getId(), PendingCpuCores.class)
+                    .stream()
+                    .filter(pending -> !pending.getCpuPinningPolicy().isExclusivelyPinned())
+                    .mapToLong(pending -> pending.getCoreCount())
+                    .max().orElse(0);
+
+            long maxRunningSharedCpuCount = resourceManager.getVdsManager(host.getId()).getMinRequiredSharedCpusCount();
+
+            long maxSharedCpuCount = LongStream.of(
+                    maxVmGroupSharedCpuCount,
+                    maxPendingSharedCount,
+                    maxRunningSharedCpuCount).max().orElse(0);
+
+            if (hostCpuCount - unavailableCpus - addedExclusivelyPinnedCpus - maxSharedCpuCount < 0) {
```

Comment:
Does `addedExclusivelyPinnedCpus` consider the `getCountThreadsAsCores()`?

Reply:
No. If `getCountThreadsAsCores() == true`, the calculation is correct as it is. if `getCountThreadsAsCores() == false`, the calculation would count the CPUs as if the dedicated policy took the whole core for each vCPU. It is because we do not know how the dedicated CPUs will be pinned to physical CPU. 

I think it makes sure we have enough CPUs for all cases even though after the pinning, we might end up with more free CPUs than we originally calculated (because the dedicated would be pinned to threads on the same core).

Reply:
But the default is `false`, I we basically support `false`, not sure if we really support `true` case.

Reply:
It's better to handle it as `getCountThreadsAsCores() == false`. Because it is the default. In our allocation logic and so on, we acted as it is `false`.

For example in this policy:
```
int futureCpus = context.getCluster().getCountThreadsAsCores() ? cores - vds.getVmsCoresCount() :
                            cores - (int) Math.ceil(vds.getVmsCoresCount() / (vds.getCpuThreads() / vds.getCpuCores()));
```
and
```
takenCpus = context.getCluster().getCountThreadsAsCores() ?
                            vdsCpuUnitPinningHelper.getDedicatedCount(vds.getId()) :
                            vdsCpuUnitPinningHelper.countTakenCores(vds);
```

In the allocation we also don't make 1:1:2 topology as 1:2:1. So I generally think we need to check it.
All of it because `hostCpuCount` *is* based on the `getCountThreadsAsCores()`. It is either CPUs or cores.

Reply:
Does it mean that the `vm.getThreadsPerCpu() <= host.getCpuCores() / host.getCpuThreads()` has to be always true? If so, we should make it explicit and maybe even introduce a new filtering policy for CPU topology as we already check the host's CPU topology for the resize and pin. Is that true also for the number of sockets and number of cores?

If we have such check, we may count the added exclusive pinning count as follows:
```
int addedExclusivelyPinnedCpus = vmGroup.stream()
    .filter(vm -> vm.getCpuPinningPolicy().isExclusive())
    .mapToInt(vm -> context.getCluster().getCountThreadsAsCores() 
        ? vm.getNumOfCpus())
        : vm.getNumOfCpus() / vm.getThreadsPerCpu() // all threads from virtual core should fit into a single physical core
    .sum();
```
Is that what you had in mind?

Reply:
> Does it mean that the `vm.getThreadsPerCpu() <= host.getCpuCores() / host.getCpuThreads()` has to be always true? 

No, I don't think it always has to be true. Let's say we don't have any policy on the VM, it can use share pool and the threads can split into multiple cores and sockets. I don't think we care about the number of threads per core.

>If so, we should make it explicit and maybe even introduce a new filtering policy for CPU topology as we already check the host's CPU topology for the resize and pin. Is that true also for the number of sockets and number of cores?
> 

We check the host's CPU topology for the resize and pin only because the way we set the VM topology which is 1:1 to the host, but with -1 core. If the core is 1, we set the VM with 0 cores (so, 0 vCPUs in total). It means it doesn't fit the policy.
> If we have such check, we may count the added exclusive pinning count as follows:
> 
> ```
> int addedExclusivelyPinnedCpus = vmGroup.stream()
>     .filter(vm -> vm.getCpuPinningPolicy().isExclusive())
>     .mapToInt(vm -> context.getCluster().getCountThreadsAsCores() 
>         ? vm.getNumOfCpus())
>         : vm.getNumOfCpus() / vm.getThreadsPerCpu() // all threads from virtual core should fit into a single physical core
>     .sum();
> ```
> 
> Is that what you had in mind?

Yes, it's close to what I had in mind. But maybe it would be easier with: `vm.getNumOfCpus(context.getCluster().getCountThreadsAsCores())` ? But, I'm also wondering if we care about the VM cores. I think we might do it with dividing to the host cores, which is what we really check here.
So that would be: vm.getNumOfCpus() / threads-per-core of the host. 
=> `vm.getNumOfCpus() / (vds.getCpuThreads() / vds.getCpuCores())` ?

Reply:
Your comment got me the idea that we may need to consider both, vm's threads and host's threads to see, how many VM's core can fit into a host core. This would be done by dividing them, eg.g for VM 3:3:3 and Host 1:9:6 we would fit 2 vCores to one pCore
```
(vds.getCpuThreads() / vds.getCpuCores()) / vm.getThreadsPerCpu()
```
Then we would divide the number of VM's cores by the vCore to pCore ratio:
```
getNumOfSockets() * getCpuPerSocket() / vCorePCoreRatio

```
For the previous topology that would be 3 * 3 / 2 = 4.5 => 5 pCores needed to accomodate this VM.

Reply:
e.g for VM 3:3:3 and Host 1:9:6 we would fit 2 vCores to one pCore.
Who said we will? It's up to the kernel and we don't know anything about it (unless it's pinned by the engine).
I think we should check the number of CPUs divided only by the threads-per-core of the host, without checking anything about the VM threads topology or anything about the topology overall.

Reply:
We are trying to count how many physical CPUs will be taken by the VM with dedicated CPU policy (this patch does not work with isolated threads yet). 

Please, consider the following example: VM 1:3:3, Host 1:3:5

if we used only host threads per core count: 
(VM has 9 CPUs, host 5 threads, 9 / 5 = we should be good with only 2 pCores)

But that would be possible only if we used the following pinning
VM's cores 0, 1, 2, 3, 4 pinned to host core 1
VM's cores 5, 6, 7, 8 pinned to host core 2

IIUC, the dedicated policy would not allow such pinning, instead, it would create the following:
VM's cores 0, 1, 2 pinned to host core 1
VM's cores 3, 4, 5 pinned to host core 2
VM's cores 6, 7, 8 pinned to host core 3

In that case, this would be more correct calculation
VM has 3 cores / (5 host threads / 3 vm threads) = we need 3 pCores

I am not still sure if I understand how the dedicated pinning is created so please correct me if I am wrong.

Reply:
about the pinning logic:
it will preserve the VM topology. the only way it won't is if we can put more vSockets within pSockets.
for example if we have a VM with: 2:1:2, the host has: 2:2:2, the actual topology of the VM would be(in terms of what physically used): 1:2:2.

why not using the existing logic?
```
takenCpus = context.getCluster().getCountThreadsAsCores() ?
                            vdsCpuUnitPinningHelper.getDedicatedCount(vds.getId()) :
                            vdsCpuUnitPinningHelper.countTakenCores(vds);
```
It will give you the number of CPUs *or* the number of taken cores based on the `getCountThreadsAsCores()`.

Reply:
I do use the current logic for counting available CPUs (I've just created a new method `countUnavailableCpus` in `VdsCpuUnitPinningHelper` that contains the following code to make the `CPUPolicyUnit` simpler
```
context.getCluster().getCountThreadsAsCores() ?
                            vdsCpuUnitPinningHelper.getDedicatedCount(vds.getId()) :
                            vdsCpuUnitPinningHelper.countTakenCores(vds);
```

However, I need to also count how many exclusively pinned CPUs will the currently scheduled vm group take (`addedExclusivelyPinnedCpus`). 

From the whole patch, this is the crucial part:
```
hostCpuCount - unavailableCpus - addedExclusivelyPinnedCpus - maxSharedCpuCount
```
And we should discuss how those are calculated:
`hostCpuCount` - the total number of host's CPU. I think we are good here as we use previous logic
`unavailableCpus` - how many CPUs are unavailable due to the existing exclusive pinning (this uses current logic from `VdsCpuUnitPinningHelper`)
`addedExclusivelyPinnedCpus` - how many CPUs will be taken by the VMs from the current VM group with exclusive pinning - that is not 100% clear
`maxSharedCpuCount` - how many shared CPUs need to remain on the host to accomodate the VM with the biggest CPU count

Reply:
This whole discussion is because `hostCpuCount`.
`Integer hostCpuCount = SlaValidator.getEffectiveCpuCores(host, context.getCluster().getCountThreadsAsCores());`
It can be either CPUs or cores (by default it will be cores).

Reply:
I do not agree, this discussion started by the following comment:
> Does `addedExclusivelyPinnedCpus` consider the `getCountThreadsAsCores()`?

I agree that the considering of the countThreadsAsCores makes things difficult. It is not necessary to consider it in every calculation though.
`hostCpuCount` - the `countThreadsAsCores` is considered here by `SlaValidator.getEffectiveCpuCores`
unavailableCpus -  the `countThreadsAsCores` is considered by `VdsCpuUnitPinningHelper.unavailableCpus()`
addedExclusivelyPinnedCpus - under discussion
maxSharedCpuCount - no need to consider it

Reply:
Yes, you are right but it's a combination of the two.
As you wrote (and I) - `hostCpuCount` can be either CPUs or cores.
Then, removing from `hostCpuCount`, if it's CPUs we need `addedExclusivelyPinnedCpus` as CPUs, if it's cores i think we need to remove cores. Am I wrong?

Reply:
You are right and I tried to address it with the following calculation 
```
int addedExclusivelyPinnedCpus = vmGroup.stream()
    .filter(vm -> vm.getCpuPinningPolicy().isExclusive())
    .mapToInt(vm -> context.getCluster().getCountThreadsAsCores() 
        ? vm.getNumOfCpus())
        : vm.getNumOfCpus() / vm.getThreadsPerCpu() // all threads from virtual core should fit into a single physical core
    .sum();
```

But in the light of our discussion about actually applying the pinning in `CpuPinningPolicyUnit` I think we should do the same here. We should apply the pinning for the whole vmGroup and then count either used threads or used cores (similar to the current `unavailableCpus`). 

```
hostCpuCount - unavailableCpus  - maxSharedCpuCount

```